[TextureMapper] Preserve-3d layers don't get flattened correctly

zhani opened the pull request:
#### 7f89d2259aab7f09b486a956fbdc7d1162fb0f92
<pre>
[TextureMapper] Preserve-3d layers don&apos;t get flattened correctly
<a href="https://bugs.webkit.org/show_bug.cgi?id=281079">https://bugs.webkit.org/show_bug.cgi?id=281079</a>

Reviewed by Nikolas Zimmermann and Fujii Hironori.

The TextureMapper incorrectly flattens layers to wrong 2D plane
in the 3D rendering context. Instead of accumulating 3D transformations
for each layer, it resets transformations to the page root layer&apos;s 2D plane
for leaf layers. This fix ensures that proper 3D transforms are applied
throughout the entire 3D context and leaf layers are flattened correctly.

Expected Behavior (as per CSS spec):

&quot;The element establishing the 3D rendering context, along with other 3D transformed
elements participating in the context, should be rendered into its own plane.
This plane includes the element’s backgrounds, borders, box decorations, content,
and descendant elements—excluding descendants that have their own planes.&quot;

* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
(WebCore::TextureMapperFlattenedLayer::TextureMapperFlattenedLayer):
(WebCore::TextureMapperFlattenedLayer::bind):
(WebCore::TextureMapperFlattenedLayer::layerRect const):
(WebCore::TextureMapperFlattenedLayer::surface const):
(WebCore::TextureMapperFlattenedLayer::paintToTextureMapper):
(WebCore::TextureMapperLayer::preprocess):
(WebCore::TextureMapperLayer::collectLayersToFlattenRecursive):
(WebCore::TextureMapperLayer::flatten):
(WebCore::TextureMapperLayer::computeFlattenedRegion):
(WebCore::TextureMapperLayer::postprocess):
(WebCore::TextureMapperLayer::freeFlattenedLayersRecursive):
(WebCore::TextureMapperLayer::computeTransformsRecursive):
(WebCore::TextureMapperLayer::paint):
(WebCore::TextureMapperLayer::paintSelf):
(WebCore::TextureMapperLayer::paintSelfAndChildren):
(WebCore::TextureMapperLayer::flattensAsLeafOf3DSceneOr3DPerspective const):
(WebCore::TextureMapperLayer::has3DLocalTransform const):
(WebCore::TextureMapperLayer::accumulatesTransform const):
(WebCore::TextureMapperLayer::computeOverlapRegions):
(WebCore::TextureMapperLayer::paintRecursive):
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:

Canonical link: <a href="https://commits.webkit.org/285820@main">https://commits.webkit.org/285820@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f629e0b8c5ca26c87883edbf42740e73dc69411e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73816 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53245 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26627 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78140 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25054 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75933 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62378 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1030 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58066 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16426 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76883 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48199 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63514 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38462 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44995 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21003 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23387 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66562 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21350 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79705 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1133 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/577 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66398 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1276 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63526 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65677 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16270 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9545 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/7808 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1097 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/3847 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1126 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/1113 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1132 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -63,6 +63,40 @@ struct TextureMapperLayer::ComputeTransformData {
 
+class TextureMapperFlattenedLayer : public TextureMapperPlatformLayer {
+public:
+    TextureMapperFlattenedLayer(const IntRect& rect)
```
```suggestion
    explicit TextureMapperFlattenedLayer(const IntRect& rect)
```

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -63,6 +63,40 @@ struct TextureMapperLayer::ComputeTransformData {
+        return FloatRect(m_rect);
+    }
+
+    RefPtr<BitmapTexture> surface() const
```
Why return a RefPtr<> here? That's unnecessary refcount churn, if you just want to access the surface and non-null is guaranteed, I'd suggest to expose 'BitmapTexture& surface() const { return *m_texture.get(); }'.

zhani replied:
Patch uses SetForScope helper to temporary switch current TextureMapperPaintOptions target surface which is type of RefPtr<>. SetForScope requires types to match

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -63,6 +63,40 @@ struct TextureMapperLayer::ComputeTransformData {
+
+    inline FloatRect layerRect() const
+    {
+        return FloatRect(m_rect);
```
Why the explicit conversion? Can't you just expose the `IntRect` you stored?

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -63,6 +63,40 @@ struct TextureMapperLayer::ComputeTransformData {
+    }
+
+private:
+
```
Superfluous newline.

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -76,6 +110,112 @@ TextureMapperLayer::~TextureMapperLayer()
+    // Reset transformations for flattening root
+    SetForScope scopedLocalTransform(m_layerTransforms.localTransform, TransformationMatrix());
+    SetForScope scopedFutureLocalTransform(m_layerTransforms.futureLocalTransform, TransformationMatrix());
+    m_layerTransforms.combined = TransformationMatrix();
```
= {} ?

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -76,6 +110,112 @@ TextureMapperLayer::~TextureMapperLayer()
+
+    // Flatten the layer
+    if (m_flattenedLayer != nullptr) {
+        delete m_flattenedLayer;
```
No manual memory management in WebKit code.

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -76,6 +110,112 @@ TextureMapperLayer::~TextureMapperLayer()
+        child->computeTransformsRecursive(data);
+
+    // Flatten the layer
+    if (m_flattenedLayer != nullptr) {
```
`s/!= nullptr//`

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -76,6 +110,112 @@ TextureMapperLayer::~TextureMapperLayer()
+    computeFlattenedRegion(layerRegion);
+
+    auto flattenedLayerRect = layerRegion.bounds();
+    auto flattenedLayer = new TextureMapperFlattenedLayer(flattenedLayerRect);
```
make_unique?

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -76,6 +110,112 @@ TextureMapperLayer::~TextureMapperLayer()
     removeFromParent();
 }
 
+void TextureMapperLayer::preprocess(TextureMapperPaintOptions& options)
```
Not clear from the naming what this function does.
Something more descriptive should be used, maybe `prepareLayerForPainting` (not good, just as first suggestion), `flatternLayerBeforePaintingIfNecessary` ?

zhani replied:
Function name updated to flattenDescendantLayersIfNecessary

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -76,6 +110,112 @@ TextureMapperLayer::~TextureMapperLayer()
+    }
+}
+
+void TextureMapperLayer::flatten(TextureMapperPaintOptions& options)
```
Also flatten is a quite special term, I'd try to use more descriptive names. One can 'flatten' a 3D transform, but not a 'layer' per-se, no?

zhani replied:
Flattening is common term used in WebKit layer compositing

zhani replied:
Term is also used in Core Animation CALayers. Also Chromium uses that term in composition

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -76,6 +110,112 @@ TextureMapperLayer::~TextureMapperLayer()
+
+void TextureMapperLayer::flatten(TextureMapperPaintOptions& options)
+{
+    // Reset transformations for flattening root
```
I see that, but why? :-) That should rather be explained.

zhani replied:
Updated comment

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -76,6 +110,112 @@ TextureMapperLayer::~TextureMapperLayer()
+
+void TextureMapperLayer::computeFlattenedRegion(Region& region)
+{
+    FloatRect rect = layerRect();
```
s/FloatRect/auto/

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -76,6 +110,112 @@ TextureMapperLayer::~TextureMapperLayer()
+    }
+}
+
+void TextureMapperLayer::postprocess()
```
Why divide this in postprocess / freeFlattenedLayersRecursive?

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -301,7 +458,7 @@ void TextureMapperLayer::sortByZOrder(Vector<TextureMapperLayer* >& array)
 }
 
-void TextureMapperLayer::paintSelfAndChildren(TextureMapperPaintOptions& options)
+void TextureMapperLayer::paintSelfAndChildren(TextureMapperPaintOptions& options, bool includesFlattenedChildren)
```
If it's generally useful, I'd rather add this to the TextureMapperPaintOptions, if not, I'd add an `enum class ShouldIncludeFlattenedChildren : bool { Yes, No }` to the header, and use that instead of a boolean - which is always harder to read on the call site, as it's not descriptive.

zhani replied:
the context for includesFlattenedChildren is only within this specific function so I prefer to keep the change simple

zhani replied:
Refactored code and this parameter is now removed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -343,8 +500,20 @@ void TextureMapperLayer::paintSelfAndChildren(TextureMapperPaintOptions& options
-        child->paintRecursive(options);
+    for (auto* child : m_children) {
+        // During flattening, already flattened children should not be painted
+        // because they are roots of their own render pass
```
What is a render pass? I don't really understand the comment.

zhani replied:
It's a stage in overall rendering were part of the scene is render into intermediate buffer. Once all passes are generated then final scene can be produced. Here we skip painting of children which are already part of flattened buffers. Comment updated

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -343,8 +500,20 @@ void TextureMapperLayer::paintSelfAndChildren(TextureMapperPaintOptions& options
+            continue;
+
+        if (isFlattened() && includesFlattenedChildren) {
+            // Only flattened child should be included from flattened layer's children.
```
I don't get that sentence.

zhani replied:
this code has been removed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -358,6 +527,38 @@ bool TextureMapperLayer::shouldBlend() const
+
+bool TextureMapperLayer::has3DLocalTransform() const
+{
+    return !(!m_layerTransforms.localTransform.m13() && !m_layerTransforms.localTransform.m23()
```
There are no existing helpers for that? I doubt it.

zhani replied:
RenderLayer has function called has3DTransform() which returns !m_transformation->isAffine(). Using TransformationMatrix::isAffine() here also probably works but that's not entirely correct because not affine doesn't mean that transformation is 3D. 3D tranformation can also be affine transformation

nikolaszimmermann replied:
I'd then add a new helper to `TransformationMatrix`  directly.

There is already:

```
    enum class Type : uint8_t {
        IdentityOrTranslation,
        Affine, 
        Other
    };
    
    Type type() const
    {
        if (!m13() && !m14() && !m23() && !m24() && !m34() && !m31() && !m32() && m33() == 1 && m44() == 1) {
            if (!m12() && !m21() && m11() == 1 && m22() == 1)
                return Type::IdentityOrTranslation;
            if (!m43()) 
                return Type::Affine;
        }
        return Type::Other;
    }
```

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -358,6 +527,38 @@ bool TextureMapperLayer::shouldBlend() const
+bool TextureMapperLayer::flattensAsLeafOf3DSceneOr3DPerspective() const
+{
+    // Check if 3D scene leaf or if has perspective and has children and has 3D transformation
+    if (((!m_state.preserves3D && (m_parent && m_parent->preserves3D())) || m_layerTransforms.localTransform.hasPerspective())
```
Does that code exist in a similar way in GraphicsLayerCA & co?

zhani replied:
nope, CALayer does flattening internally in Core Animation based on given layer properties

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -358,6 +527,38 @@ bool TextureMapperLayer::shouldBlend() const
+    return !m_state.masksToBounds && (preserves3D() || (m_parent && m_parent->preserves3D()));
+}
+
+bool TextureMapperLayer::preserves3D() const
```
Candidate for inlining in the header.

zhani replied:
Inlining causes style check failure "Inline functions should not be in classes annotated with WEBCORE_EXPORT"

fujii replied:
It sholuld be inlined. The best approach in this case is removing the class-level WEBCORE_EXPORT before this change.

fujii replied:
I will do that in https://bugs.webkit.org/show_bug.cgi?id=281899

zhani replied:
ok, now I understand. Sorry for misunderstanding. Thanks for the patch

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -358,6 +527,38 @@ bool TextureMapperLayer::shouldBlend() const
+
+bool TextureMapperLayer::isFlattened() const
+{
+    return m_flattenedLayer != nullptr;
```
Candidate for inlining in the header.

zhani replied:
Inlining causes style check failure "Inline functions should not be in classes annotated with WEBCORE_EXPORT"

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:
```diff
@@ -181,6 +189,18 @@ class WEBCORE_EXPORT TextureMapperLayer : public CanMakeWeakPtr<TextureMapperLay
     bool shouldBlend() const;
 
+    bool flattensAsLeafOf3DScene() const;
+
```
Remove superfluous newlines here.

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:
```diff
@@ -191,6 +211,7 @@ class WEBCORE_EXPORT TextureMapperLayer : public CanMakeWeakPtr<TextureMapperLay
     WeakPtr<TextureMapperLayer> m_effectTarget;
     TextureMapperBackingStore* m_backingStore { nullptr };
     TextureMapperPlatformLayer* m_contentsLayer { nullptr };
+    TextureMapperFlattenedLayer* m_flattenedLayer { nullptr };
```
I'd rather use std::unique_ptr<> here, especially in new code.

zhani replied:
Fixed

nikolaszimmermann reviewed (changes requested):
Thanks @zhani! Looking forward to get this fixed.

Phew, I have a hard time understanding this patch, but I would like to :-) I hope you can improve the commit message, to me it's quite unclear _how_ you are fixing things, and _what_ the current problems are. Ideally with more examples, what is going wrong. Even better with the smallest possible testcase, that you then dissect.

fujii reviewed (changes requested):
It's inefficient to allocate TextureMapperFlattenedLayer for all flattened layers in `preprocess`.

In my WIP patch https://bugs.webkit.org/show_bug.cgi?id=241699#c5
TextureMapperLayer::paint2DRoot allocates a temporary texture only for the current 2D layer.

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -63,6 +63,40 @@ struct TextureMapperLayer::ComputeTransformData {
     }
 };
 
+class TextureMapperFlattenedLayer final : public TextureMapperPlatformLayer {
```
TextureMapperFlattenedLayer doesn't need to inherit TextureMapperPlatformLayer.

zhani replied:
Fixed

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -76,6 +110,107 @@ TextureMapperLayer::~TextureMapperLayer()
+    // Preprocess is called as a first step before painting any layers
+
+    // Layers are collected in depth first, post-order
+    Vector<TextureMapperLayer*> layersToFlatten;
```
You don't need this array. You can just render flatten layers recursively.

zhani replied:
Fixed

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -76,6 +110,107 @@ TextureMapperLayer::~TextureMapperLayer()
+
+    // Reset transformations for flattening root 2D
+    SetForScope scopedLocalTransform(m_layerTransforms.localTransform, TransformationMatrix());
+    SetForScope scopedFutureLocalTransform(m_layerTransforms.futureLocalTransform, TransformationMatrix());
```
Windows port is broken.
https://gist.github.com/fujii/dd5e25cbc0f5241a360390033c4fed84

zhani replied:
Thanks, I applied the change

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -76,6 +110,107 @@ TextureMapperLayer::~TextureMapperLayer()
+    if (m_state.backdropLayer)
+        m_state.backdropLayer->computeTransformsRecursive(data);
+    for (auto* child : m_children)
+        child->computeTransformsRecursive(data);
```
TextureMapperLayer::paint also calls computeTransformsRecursive for all layers.
It's redundant to compute transforms for a layer multiple times.
In my WIP patch, it is computed only once.

zhani replied:
For flattening 2D root, child transforms need to be calculated for that root. After flattening is done, then transforms need to be recalculated taking account flattened layer rect. This is relatively cheap operation because trees are usually not very deep and calculation involved is not complex

fujii replied:
I think this approach doesn't work with a backdrop filter.
A backdrop layer renders the root layer with identity matrix into a temporary buffer.
https://github.com/WebKit/WebKit/blob/a10ee6a9b1ee5949ced6cfdab1a393243e65a9b3/Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp#L725-L729
`computeTransformsRecursive` should be done before it.

zhani replied:
You right, this seems to fail. I'm working on a fix

zhani replied:
Backdrops are now working according to spec. I added new test case
https://bug-281079-attachments.webkit.org/attachment.cgi?id=473058

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -76,6 +110,107 @@ TextureMapperLayer::~TextureMapperLayer()
+    // Flattens descendant layers to into this planes z=0. In theory we could
+    // project each sublayer to this layer's plane but in practice that doesn't
+    // result in pixel perfect outcome and that's why each sublayer needs to be
+    // rendered into this layer's in 2D root.
```
I don't think this comment is true.  In my understanding, all descendants of the flatten layer have to be rendered to z=0 of the flatten layer.

zhani replied:
What I tried to explain here is that instead of flattening subtree into separate 2D buffer we could try projecting child layers to flattening layer's plane in 3D coordinate space and thus avoid separate flattening buffer but in practice that does not result in pixel perfect outcome due to precision issues. I removed the comment to avoid confusion

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -358,6 +520,38 @@ bool TextureMapperLayer::shouldBlend() const
+
+bool TextureMapperLayer::accumulatesTransformForChildren() const
+{
+    return preserves3D() || m_layerTransforms.localTransform.hasPerspective();
```
Why do you need `hasPerspective` here. Do you have a test case?

zhani replied:
Not needed. Removed

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -80,6 +114,92 @@ TextureMapperLayer::~TextureMapperLayer()
+    computeFlattenedRegion(layerRegion);
+
+    auto flattenedLayerRect = layerRegion.bounds();
+    auto flattenedLayer = WTF::makeUnique<TextureMapperFlattenedLayer>(flattenedLayerRect);
```
Why don't reuse m_flattenedLayer? a single layer is rendered multiple times.

fujii replied:
I misunderstand. Never mind.

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -67,6 +67,40 @@ struct TextureMapperLayer::ComputeTransformData {
+    }
+
+private:
+    IntRect m_rect;
```
BitmapTexture has size. I think you can remove m_rect.

zhani replied:
flattened layer may exceed the bounds of host layer in every direction not only size but x,y position also

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -80,6 +114,92 @@ TextureMapperLayer::~TextureMapperLayer()
+    auto flattenedLayerRect = layerRegion.bounds();
+    auto flattenedLayer = WTF::makeUnique<TextureMapperFlattenedLayer>(flattenedLayerRect);
+
+    flattenedLayer->bind(options.textureMapper);
```
You have to take `maxTextureSize` into account.
My WIP patch did it. https://bugs.webkit.org/show_bug.cgi?id=241699#c5

zhani replied:
Fixed

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -80,6 +114,92 @@ TextureMapperLayer::~TextureMapperLayer()
 
+void TextureMapperLayer::flattenDescendantLayersIfNecessary(TextureMapperPaintOptions& options)
+{
+    // Preprocess is called as a first step before painting any layers
```
Remove this comment.

zhani replied:
Fixed

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -80,6 +114,92 @@ TextureMapperLayer::~TextureMapperLayer()
+    if (m_state.backdropLayer)
+        m_state.backdropLayer->computeTransformsRecursive(data);
+    for (auto* child : m_children)
+        child->computeTransformsRecursive(data);
```
`data` is unused. Do this:
> option.textureMapper.setDepthRange(data.zNear, data.zFar);

zhani replied:
Fixed

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -362,6 +499,39 @@ bool TextureMapperLayer::shouldBlend() const
+{
+    bool isLeafOf3DScene = !m_state.preserves3D && (m_parent && m_parent->preserves3D());
+    bool hasPerspective = m_layerTransforms.localTransform.hasPerspective() && m_layerTransforms.localTransform.m34() != -1.f;
+    if ((isLeafOf3DScene || hasPerspective) && !m_children.isEmpty() && has3DLocalTransform())
```
In my understanding, the condtion should be `if (isLeafOf3DScene && !m_children.isEmpty())`.

zhani replied:
No, there's not need to flatten to intermediate buffer if there's no 3D transform in leaf node. In case no 3D leaf can be rendered to parent surface. Also perspective with 3D transform need to propagate perspective and 3D transform to all of its child layers so it behaves in a similar manner as 3D leaf. There's test scenario for perspective case in https://bugs.webkit.org/show_bug.cgi?id=281113

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -362,6 +499,39 @@ bool TextureMapperLayer::shouldBlend() const
+}
+
+bool TextureMapperLayer::accumulatesTransformForChildren() const
+{
```
This method looks very useless. Why do you want to add this?

zhani replied:
Originally this had other conditions which are now obsolete. Now this function is removed

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -80,6 +114,92 @@ TextureMapperLayer::~TextureMapperLayer()
+    options.textureMapper.bindSurface(options.surface.get());
+}
+
+void TextureMapperLayer::computeFlattenedRegion(Region& region)
```
This code doesn't take a replica layer into account.
In my WIP patch, computeOverlapRegions doesn't it.
And, it takes clip rect too.

zhani replied:
I've made the necessary adjustments to account for the replica layer. General clipping within flattened layers is already handled, but the patch was missing the case where the flattened layer needs to be clipped against the viewport rect. This situation is quite rare, and even the WIP patch doesn’t fully handle cases with nested flattenings. To properly address viewport clipping in nested flattenings, the logic for tracking and calculating the accumulated transform would need to be reworked to avoid redundant calculations. Also, with this patch, viewport clipping does not fail, but in the worst case, it may allocate an unnecessarily large intermediate flattening buffer. I was thinking of addressing this in a separate patch, as this one is already quite large and covers several use cases.

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -67,6 +67,71 @@ struct TextureMapperLayer::ComputeTransformData {
+    void paintToTextureMapper(TextureMapper& textureMapper, const FloatRect& targetRect, TransformationMatrix& modelViewMatrix, float opacity)
+    {
+        int textureIndex = 0;
+        IntSize maxTextureSize = textureMapper.maxTextureSize();
```
I think `allEdgesExposed` should be calcurated here, the outside of loops.

~~~
  bool allEdgesExposed = m_rect.width() <= maxTextureSize.width() && m_rect.height() <= maxTextureSize.height() ;
~~~

zhani replied:
Fixed

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -67,6 +67,71 @@ struct TextureMapperLayer::ComputeTransformData {
+        IntSize maxTextureSize = textureMapper.maxTextureSize();
+        for (int x = m_rect.x(); x < m_rect.maxX(); x += maxTextureSize.width()) {
+            for (int y = m_rect.y(); y < m_rect.maxY(); y += maxTextureSize.height()) {
+                IntRect tileRect(IntPoint(x, y), maxTextureSize);
```
I think we use `{ ... }` in this case. 
~~~
 IntRect tileRect({ x, y }, maxTextureSize);
~~~

zhani replied:
Fixed

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -67,6 +67,71 @@ struct TextureMapperLayer::ComputeTransformData {
+                    options.textureMapper.bindSurface(options.surface.get());
+                    paintFunction(options);
+                }
+                m_textures.append(surface.ptr());
```
~~~
 m_textures.append(WTFMove(surface));
~~~

zhani replied:
Fixed

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:
```diff
@@ -188,6 +200,7 @@ class WEBCORE_EXPORT TextureMapperLayer : public CanMakeWeakPtr<TextureMapperLay
     WeakPtr<TextureMapperLayer> m_effectTarget;
     TextureMapperBackingStore* m_backingStore { nullptr };
     TextureMapperPlatformLayer* m_contentsLayer { nullptr };
+    std::unique_ptr<TextureMapperFlattenedLayer> m_flattenedLayer { nullptr };
```
We don't need `{ nullptr }` here.

zhani replied:
Fixed

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -129,7 +268,7 @@ void TextureMapperLayer::computeTransformsRecursive(ComputeTransformData& data)
+            m_layerTransforms.futureCombinedForChildren.flatten();
         m_layerTransforms.futureCombinedForChildren.multiply(m_state.childrenTransform);
         m_layerTransforms.futureCombinedForChildren.translate3d(-originX, -originY, -m_state.anchorPoint.z());
 #endif
```
I think `computeTransformsRecursive` should return early if `isFlattened()` because it won't render the child layers.

zhani replied:
Fixed. Doesn't recurse computation to children if already flattened

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -67,6 +67,63 @@ struct TextureMapperLayer::ComputeTransformData {
+                tileRect.intersect(m_rect);
+
+                TransformationMatrix adjustedTransform = modelViewMatrix * TransformationMatrix::rectToRect(m_rect, targetRect);
+                textureMapper.drawTexture(*m_textures[textureIndex++].get(), tileRect, adjustedTransform, opacity, allEdgesExposed ? TextureMapper::AllEdgesExposed::Yes : TextureMapper::AllEdgesExposed::No);
```
Move this code `allEdgesExposed ? TextureMapper::AllEdgesExposed::Yes : TextureMapper::AllEdgesExposed::No` to the outside of this loop.

zhani replied:
Fixed

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -67,6 +67,79 @@ struct TextureMapperLayer::ComputeTransformData {
+                    options.textureMapper.bindSurface(options.surface.get());
+                    paintFunction(options);
+
+                    surface = Ref { *options.surface };
```
Why do you need this line?

zhani replied:
It's for the case when the element which initiates the flattening layer has a filter for example drop shadow. When TextureMapper applies the effect it replaces the original surface and sets new surface to options.surface. The way how filters work for flattened layer is that the filter is applied to whole layer. Thus we need to render this new surface texture that filter assigned instead the original flattened surface texture.
I added a test case for this https://bug-281079-attachments.webkit.org/attachment.cgi?id=473059

fujii replied:
I got it. paintIntoSurface create another BitmapTexture. Never mind.

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -347,8 +519,10 @@ void TextureMapperLayer::paintSelfAndChildren(TextureMapperPaintOptions& options
 
-    for (auto* child : m_children)
-        child->paintRecursive(options);
+    if (!isFlattened() || (isFlattened() && !m_flattenedLayer->valid())) {
```
This condition can be simplified.
> if (!isFlattened() || !m_flattenedLayer->valid()) {

zhani replied:
Good catch. Fixed

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -80,6 +153,87 @@ TextureMapperLayer::~TextureMapperLayer()
+
+    Region layerRegion;
+    computeFlattenedRegion(layerRegion, true);
+
```
I think we can remove `TextureMapperFlattenedLayer::valid()` if we return early here.
~~~c++
if (layerRegion.isEmpty())
  return;
~~~

fujii replied:
I misunderstood. Never mind.

zhani replied:
`valid()` was a bit bad choice for function name. I renamed it to `needsUpdate().` Also added your suggestion for `if (layerRegion.isEmpty())`

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -809,14 +1004,40 @@ void TextureMapperLayer::paintRecursive(TextureMapperPaintOptions& options)
+    if (m_state.backdropLayer && !options.backdropLayer)
+        paintBackdrop(options);
+
+    if (m_flattenedLayer && !m_flattenedLayer->valid()) {
```
m_flattenedLayer is not needed here.

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -67,6 +67,79 @@ struct TextureMapperLayer::ComputeTransformData {
     }
 };
 
+class TextureMapperFlattenedLayer final {
```
I would prefer this to move into its own .cpp/h file -- this is already large enough.

zhani replied:
I'm planning to do some refactoring to unify tiled updates which are also used in other places in very similar manner. Maybe it's better to move this to its own file in separate patch if the class still grows in size. Now it's quite small and only for internal usage of TextureMapperLayer.

nikolaszimmermann replied:
Still, we have a general policy of one class per .cpp/h file in WebKit, which was already violated, and this doesn't make it better :-)

zhani replied:
TextureMapperFlattenedLayer uses TextureMapperPaintOptions which is also internal to TextureMapperLayer. TextureMapperPaintOptions  is a container to hold values during painting internally and it needs to be moved out of .cpp also if TextureMapperFlattenedLayer is moved. I still think that maybe it's better to do that in another patch because there are other refactoring needed also.

nikolaszimmermann replied:
OK, usually I tend to do this kind of refactoring before in a separate PR. But we can also postpone the cleanup to a follow-up PR. I don't want to block this important work.

zhani replied:
ok, I'll do follow up patch. Thanks

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -67,6 +67,79 @@ struct TextureMapperLayer::ComputeTransformData {
+                    options.textureMapper.bindSurface(options.surface.get());
+                    paintFunction(options);
+
+                    surface = Ref { *options.surface };
```
Still this kind of memory managment seems error-prone. Easy to forget to ref.
Also a few lines before where you have to set options.surface to 'surface.ptr()' is fragile. I would suggest to rework this part.

zhani replied:
options.surface might be different surface than what is binded before paintFunction call in case if filters are applied and new filtered surface is created (such as dropshadow). TextureMapper::applyFilters creates this new surface and places it into options.surface member. Unless we don't take ref to it then it's gone after we go out of scope

zhani replied:
Fixed this. acquireTextureFromPool returns Ref<BitmapTexture> but we can assign it to RefPtr locally and then we can use options.surface without explicit Ref

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -67,6 +67,79 @@ struct TextureMapperLayer::ComputeTransformData {
+                IntRect tileRect({ x, y }, maxTextureSize);
+                tileRect.intersect(m_rect);
+
+                TransformationMatrix adjustedTransform = modelViewMatrix * TransformationMatrix::rectToRect(m_rect, targetRect);
```
Why is this done in the for-loop? Neither modelViewMatrix nor m_rect or targetRect changes?

zhani replied:
Good point. Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -67,6 +67,79 @@ struct TextureMapperLayer::ComputeTransformData {
+
+    void paintToTextureMapper(TextureMapper& textureMapper, const FloatRect& targetRect, TransformationMatrix& modelViewMatrix, float opacity)
+    {
+        int textureIndex = 0;
```
A counter can be `unsigned`.

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -67,6 +67,79 @@ struct TextureMapperLayer::ComputeTransformData {
+    }
+
+private:
+
```
You can omit the newline.

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -67,6 +67,79 @@ struct TextureMapperLayer::ComputeTransformData {
+                tileRect.intersect(m_rect);
+
+                TransformationMatrix adjustedTransform = modelViewMatrix * TransformationMatrix::rectToRect(m_rect, targetRect);
+                textureMapper.drawTexture(*m_textures[textureIndex++].get(), tileRect, adjustedTransform, opacity, allEdgesExposed);
```
IMHO it's dangerous to rely that the textureIndex here is valid. When any of the both two outer for loops here or in `update` are changed, this breaks. It would be better to refactor the logic into a single helper function, that centralizes the x/y paritioning <-> texture mapping logic and is shared between `update` and `paintToTextureMapper`. Note: I dislike the `update` name a bit, it's quite generic, and one doesn't really know what's happening just from reading the function name.

zhani replied:
Fixed. Loop logic is now moved to helper function

nikolaszimmermann replied:
How about calling it forEachTile? 😂

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -80,6 +153,90 @@ TextureMapperLayer::~TextureMapperLayer()
+
+void TextureMapperLayer::processFlatteningRequirements()
+{
+    if (m_flattenedLayer)
```
No need to null check, prio to restting to `nullptr`.

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -80,6 +153,90 @@ TextureMapperLayer::~TextureMapperLayer()
+    SetForScope scopedLocalTransform(m_layerTransforms.localTransform, TransformationMatrix());
+    m_layerTransforms.combined = { };
+    m_layerTransforms.combinedForChildren = { };
+#if USE(COORDINATED_GRAPHICS)
```
Why is only this part protected by the guard?

zhani replied:
Windows port also uses TextureMapper but does not use COORDINATED_GRAPHICS.
future* transforms are specific to COORDINATED_GRAPHICS

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -80,6 +153,90 @@ TextureMapperLayer::~TextureMapperLayer()
+        m_flattenedLayer = nullptr;
+
+    // Reset transformations as this layer is root 2D
+    SetForScope scopedLocalTransform(m_layerTransforms.localTransform, TransformationMatrix());
```
Use `TransformationMatrix::identity`.

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -80,6 +153,90 @@ TextureMapperLayer::~TextureMapperLayer()
+    m_layerTransforms.combined = { };
+    m_layerTransforms.combinedForChildren = { };
+#if USE(COORDINATED_GRAPHICS)
+    SetForScope scopedFutureLocalTransform(m_layerTransforms.futureLocalTransform, TransformationMatrix());
```
Use `TransformationMatrix::identity`.

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -80,6 +153,90 @@ TextureMapperLayer::~TextureMapperLayer()
+    if (layerRegion.isEmpty())
+        return;
+
+    m_flattenedLayer = WTF::makeUnique<TextureMapperFlattenedLayer>(layerRegion.bounds(), data.zNear, data.zFar);
```
s/WTF:://

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -80,6 +153,90 @@ TextureMapperLayer::~TextureMapperLayer()
+
+void TextureMapperLayer::freeFlattenedDescendantLayers()
+{
+    if (m_flattenedLayer)
```
Remove the if condition.

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -80,6 +153,90 @@ TextureMapperLayer::~TextureMapperLayer()
+    }
+}
+
+void TextureMapperLayer::freeFlattenedDescendantLayers()
```
s/free/destroy/ ? s/free/purge/ ? s/free/remove/ ?

zhani replied:
Renamed to destroyFlattenedDescendantLayers()

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -583,7 +778,9 @@ void TextureMapperLayer::computeOverlapRegions(ComputeOverlapRegionData& data, c
 
     FloatRect localBoundingRect;
-    if (m_backingStore || m_state.masksToBounds || m_state.maskLayer || hasFilters())
+    if (isFlattened() && !m_isBackdrop)
```
What happens in the `m_isBackdrop` case?

zhani replied:
Backdrops are applied only to owning element bounding rect not to whole flattened area

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:
```diff
@@ -171,13 +181,27 @@ class TextureMapperLayer : public CanMakeWeakPtr<TextureMapperLayer> {
+    bool flattensAsLeafOf3DSceneOr3DPerspective() const;
+    bool has3DLocalTransform() const;
+
+    inline bool preserves3D() const
```
No need to explict inline -- it is inlined already. Also you can consolidate these getters to one-liners.

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:
```diff
@@ -171,13 +181,27 @@ class TextureMapperLayer : public CanMakeWeakPtr<TextureMapperLayer> {
+
+    inline bool isFlattened() const
+    {
+        return m_flattenedLayer != nullptr;
```
`return !!m_flattenedLayer`

zhani replied:
Fixed

nikolaszimmermann reviewed (changes requested):
IMHO this should be cleaned-up, in a final step, now that @fujii is happy with the logic.

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapper.h:
```diff
@@ -83,6 +83,7 @@ class TextureMapper {
     IntRect clipBounds();
     IntSize maxTextureSize() const { return IntSize(2000, 2000); }
     void setDepthRange(double zNear, double zFar);
+    void getDepthRange(double& zNear, double& zFar);
```
Note for your cleanup follow-up:

auto [zNear, zFar] = xxx.depthRange();

Is nicer using tuples.

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -67,6 +67,91 @@ struct TextureMapperLayer::ComputeTransformData {
+        , m_zFar(zFar)
+    { }
+
+    inline IntRect layerRect() const
```
Still some superfluous inlines here (you can also reduce the simple getters to one-liners).

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -67,6 +67,91 @@ struct TextureMapperLayer::ComputeTransformData {
+
+    inline bool needsUpdate() const
+    {
+        return !m_textures.size();
```
isEmpty?

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -67,6 +67,91 @@ struct TextureMapperLayer::ComputeTransformData {
+
+        m_textures.clear();
+
+        IntSize maxTextureSize = options.textureMapper.maxTextureSize();
```
auto

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -67,6 +67,91 @@ struct TextureMapperLayer::ComputeTransformData {
+
+        IntSize maxTextureSize = options.textureMapper.maxTextureSize();
+        tileLoop(maxTextureSize, [&](const IntRect& tileRect) {
+            RefPtr<BitmapTexture> surface = options.textureMapper.acquireTextureFromPool(tileRect.size(), { BitmapTexture::Flags::SupportsAlpha });
```
auto

zhani replied:
This is on purpose. aquireTextureFromPool returns Ref<BitmapTexture> but we need RefPtr to make it work with SetForScope and TextureMapperPaintOptions which stores surface as RefPtr<BitmapTexture>. Otherwise we would need this Ref { *options.surface} line which was discussed before

nikolaszimmermann replied:
Ouch you've got me 😂 fair enough!

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -67,6 +67,91 @@ struct TextureMapperLayer::ComputeTransformData {
+
+    void paintToTextureMapper(TextureMapper& textureMapper, const FloatRect& targetRect, TransformationMatrix& modelViewMatrix, float opacity)
+    {
+        IntSize maxTextureSize = textureMapper.maxTextureSize();
```
auto

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -67,6 +67,91 @@ struct TextureMapperLayer::ComputeTransformData {
+        auto allEdgesExposed = m_rect.width() <= maxTextureSize.width() && m_rect.height() <= maxTextureSize.height()
+            ? TextureMapper::AllEdgesExposed::Yes
+            : TextureMapper::AllEdgesExposed::No;
+        TransformationMatrix adjustedTransform = modelViewMatrix;
```
You could reduce that back into one line and directly use operator *.

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -67,6 +67,91 @@ struct TextureMapperLayer::ComputeTransformData {
+
+        size_t textureIndex = 0;
+        tileLoop(maxTextureSize, [&](const IntRect& tileRect) {
+            if (textureIndex < m_textures.size())
```
I would assert that!

zhani replied:
Fixed

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -67,6 +67,91 @@ struct TextureMapperLayer::ComputeTransformData {
+
+private:
+    template <typename Function>
+    void tileLoop(const IntSize& maxTextureSize, Function&& func)
```
Use function - no abbreviations per WebKit style.

zhani replied:
Fixed

nikolaszimmermann reviewed (commented):
Nice, still a final round of comments, sorry.

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -67,6 +67,82 @@ struct TextureMapperLayer::ComputeTransformData {
+    }
+
+private:
+    void forEachTile(const IntSize& maxTextureSize, const std::function<void(const IntRect&)>& func)
```
Still says func?

zhani replied:
Fixed

nikolaszimmermann reviewed (approved):
r=me

zhani committed:
[TextureMapper] Preserve-3d layers don't get flattened correctly
https://bugs.webkit.org/show_bug.cgi?id=281079

Reviewed by Nikolas Zimmermann and Fujii Hironori.

The TextureMapper incorrectly flattens layers to wrong 2D plane
in the 3D rendering context. Instead of accumulating 3D transformations
for each layer, it resets transformations to the page root layer's 2D plane
for leaf layers. This fix ensures that proper 3D transforms are applied
throughout the entire 3D context and leaf layers are flattened correctly.

Expected Behavior (as per CSS spec):

"The element establishing the 3D rendering context, along with other 3D transformed
elements participating in the context, should be rendered into its own plane.
This plane includes the element’s backgrounds, borders, box decorations, content,
and descendant elements—excluding descendants that have their own planes."

* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
(WebCore::TextureMapperFlattenedLayer::TextureMapperFlattenedLayer):
(WebCore::TextureMapperFlattenedLayer::bind):
(WebCore::TextureMapperFlattenedLayer::layerRect const):
(WebCore::TextureMapperFlattenedLayer::surface const):
(WebCore::TextureMapperFlattenedLayer::paintToTextureMapper):
(WebCore::TextureMapperLayer::preprocess):
(WebCore::TextureMapperLayer::collectLayersToFlattenRecursive):
(WebCore::TextureMapperLayer::flatten):
(WebCore::TextureMapperLayer::computeFlattenedRegion):
(WebCore::TextureMapperLayer::postprocess):
(WebCore::TextureMapperLayer::freeFlattenedLayersRecursive):
(WebCore::TextureMapperLayer::computeTransformsRecursive):
(WebCore::TextureMapperLayer::paint):
(WebCore::TextureMapperLayer::paintSelf):
(WebCore::TextureMapperLayer::paintSelfAndChildren):
(WebCore::TextureMapperLayer::flattensAsLeafOf3DSceneOr3DPerspective const):
(WebCore::TextureMapperLayer::has3DLocalTransform const):
(WebCore::TextureMapperLayer::accumulatesTransform const):
(WebCore::TextureMapperLayer::computeOverlapRegions):
(WebCore::TextureMapperLayer::paintRecursive):
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:

Canonical link: https://commits.webkit.org/285820@main
